cmd/control_loop/controller/controller.go

```diff
@@ -0,0 +1,200 @@
+
+import (
+    "github.com/apigee/registry/cmd/control_loop/list"
+    resource_lib "github.com/apigee/registry/cmd/control_loop/resources"
+    "reflect"
+    "github.com/apigee/registry/cmd/control_loop/manifest_reader"
+    "context"
+	"github.com/apigee/registry/connection"
+	"time"
+	"fmt"
+)
+
+type ResourceCollection struct {
+	maxUpdateTime time.Time
+	resources *[]resource_lib.Resource
+}
+
+func ProcessManifest(manifest *manifest_reader.Manifest) ([]string, error) {
+
+	var actions []string
+	for _, entry := range manifest.Entries {
+    	ctx := context.TODO()
+		client, err := connection.NewClient(ctx)
+		if err != nil {
+			return nil, err
+		}		
+		resourcePattern := fmt.Sprintf("projects/%s/%s", manifest.Project, entry.Resource)
+
+		// Process Dependencies
+		dependencyMaps := make([]map[string]*ResourceCollection, 0, 0)
+		err = processDependencies(
+			resourcePattern,
+			entry.Dependencies,
+			&dependencyMaps,
+			ctx,
+			client)
+		if err != nil {
+			return nil, err
+		}
+
+		cmds, err := evaluateResources(
+			resourcePattern,
+			entry.Dependencies,
+			dependencyMaps,
+			entry.Action,
+			ctx,
+			client)
+		if err != nil {
+			return nil, err
+		}
+		if len(cmds) > 0 {
+			actions = append(actions, cmds...)
+		}
+	}
+
+	return actions, nil
+}
+
+func processDependencies(
+	resourcePattern string,
+	dependencies []manifest_reader.Dependency,
+	dependencyMaps *[]map[string]*ResourceCollection,
+	ctx context.Context,
+	client connection.Client) error {
+	for _, d := range dependencies {
+		sourceMap := make(map[string]*ResourceCollection)
+		// Extend the source pattern if it contains $resource.api like pattern
+		sourcePattern, err := ParseSourcePattern(resourcePattern, d.Source)
+		if err != nil {
+			return err
+		}
+
+		// Fetch resources using the sourcePattern
+		sourceList, err :=  list.ListResources(sourcePattern, d.Filter, ctx, client)
+		if err != nil {
+			return err
+		}
+
+		// Extract the attribute which will be used
+		// to group the contents of the sourceList into a map
+		groupFuncName := ParseGroupFunc(d.Source)
+		
+		// Build source map
+		for _, source := range sourceList {
+			// Group all the resources in one grop by default
+			groupBy := ""
+
+			if len(groupFuncName) > 0 {
+				groupFunc := reflect.ValueOf(source).MethodByName(groupFuncName)
+				groupBy = groupFunc.Call([]reflect.Value{})[0].String()
+			}
+
+			// Update the map with timestamp
+			source_ts := source.GetUpdateTimestamp()
+			if collection, ok := sourceMap[groupBy]; !ok {
+				sourceMap[groupBy] = &ResourceCollection {
+					maxUpdateTime: source_ts,
+					resources: &[]resource_lib.Resource{},
+				}
+			} else {
+				if (*collection).maxUpdateTime.Before(source_ts) {
+					(*collection).maxUpdateTime = source_ts
+				}
+			}
+
+			temp := (*sourceMap[groupBy]).resources
+			(*temp) = append((*temp), source)
+		}
+		(*dependencyMaps) = append((*dependencyMaps), sourceMap)
+
+	}
+
+	return nil
+}
+
+func evaluateResources(
```

Comment:
What do you think about calling this "updateResources"? I suggest this because I think this step is bringing the resources up-to-date with the manifest.

Reply:
Done